edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/M2Mqtt/M2MqttClient.cs

```diff
@@ -0,0 +1,273 @@
+
+            var source = new TaskCompletionSource<bool>();
+            Task<bool> unsubscribeTaskSupplier()
+            {
+                lock (_lock)
+                {
+                    if (!cancellationToken.IsCancellationRequested)
+                    {
+                        var subscribeId = _client.Unsubscribe(topics.ToArray());
+                        _pendingVoidTasks.TryAdd(subscribeId, source);
+                    }
+                }
+
+                return source.Task;
+            }
+
+            Task restoreTaskSupplier()
+            {
+                source.TrySetCanceled();
+                return Task.FromResult(true);
+            }
+
+            return RunTaskWithCancellationTokenAsync(unsubscribeTaskSupplier, restoreTaskSupplier, cancellationToken);
+        }
+
+        public void RegisterConnectionStatusListener(IConnectionStatusListener connectionStatusListener) => _connectionStatusListener = connectionStatusListener;
+
+        public void RegisterMessageHandler(IMessageHandler messageHandler) => _messageHandler = messageHandler;
+
+        public bool IsConnected() => _client.IsConnected;
+
+        void HandleDisconnection(object sender, EventArgs e) => _connectionStatusListener?.onDisconnected(this, new MqttException(message: e.ToString()));
+
+        void OnMessagePublished(object sender, MqttMsgPublishedEventArgs eventArgs)
+        {
+            lock (_lock)
+            {
+                _pendingVoidTasks.Remove(eventArgs.MessageId, out var source);
+                source?.TrySetResult(eventArgs.IsPublished);
+            }
+        }
+
+        void OnMessageReceived(object sender, MqttMsgPublishEventArgs eventArgs) => _messageHandler.ProcessMessageAsync(eventArgs.Topic, eventArgs.Message).GetAwaiter().GetResult();
+
+        void OnTopicSubscribed(object sender, MqttMsgSubscribedEventArgs eventArgs)
+        {
+            lock (_lock)
+            {
+                _pendingSubscribes.Remove(eventArgs.MessageId, out var source);
+                source?.TrySetResult(eventArgs.GrantedQoSLevels);
+            }
+        }
+
+        void OnTopicUnsubscribed(object sender, MqttMsgUnsubscribedEventArgs eventArgs)
+        {
+            lock (_lock)
+            {
+                _pendingVoidTasks.Remove(eventArgs.MessageId, out var source);
+                source?.TrySetResult(true);
+            }
+        }
+
+        async Task<T> RunTaskWithCancellationTokenAsync<T>(Func<Task<T>> taskSupplier, Func<Task> restoreTaskSupplier, CancellationToken cancellationToken)
+        {
+            var waitTask = Task.FromCanceled(cancellationToken);
+            Task finishedTask;
+            Task<T> targetTask;
+            try
+            {
+                targetTask = taskSupplier();
+                finishedTask = await Task.WhenAny(targetTask, waitTask);
+            }
+            catch (Exception e)
+            {
+                throw new MqttException(cause: e);
+            }
+
```

Comment:
If operation is cancelled before finish, `restrore` will be called to cleanup the resource. Otherwise the operation either finished or failed with error. It will rely on caller to handle the result instead of code itself.